Added integration test that tests that the first frame callback is called

gaaclarke opened the pull request:
related PR: https://github.com/flutter/engine/pull/9952

dnfield commented on shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm:
```diff
@@ -89,14 +101,19 @@
+
+  // Check other bundles if we still haven't found anything.  This is useful for XCTest targets.
+  if (settings.application_library_path.size() == 0) {
+    for (NSBundle* bundle in [NSBundle allBundles]) {
```
What happens if I have an add2app case here with multiple App.frameworks embedded in different bundles?  Can that happen?

gaaclarke replied:
This doesn't get executed unless you don't specify the dart project, and only if you didn't find "info.plist[FLTLibraryPath]" and "mainBundle/App.framework".  I could make it more specific to only look in bundles that are named ".xctest".  I don't know if it is necessary though.

dnfield replied:
I'm worried we're missing something here that this could break unexpectedly, but I can't think of what.  It's probably fine.

dnfield commented on testing/scenario_app/ios/Scenarios/Scenarios/AppDelegate.m:
```diff
@@ -5,7 +5,7 @@ @implementation AppDelegate
     didFinishLaunchingWithOptions:(NSDictionary*)launchOptions {
   self.window = [[UIWindow alloc] initWithFrame:[[UIScreen mainScreen] bounds]];
-  self.window.rootViewController = [[FlutterViewController alloc] init];
+  self.window.rootViewController = [[UIViewController alloc] init];
```
I'm not clear on why we want to do this.  Your test should still work without changing this, right?

gaaclarke replied:
Nope, I have to setup the callback before the view controller is presented otherwise it would be flakey.  We couldn't guarantee that I set the callback up before the first frame got presented.

dnfield replied:
But you're testing a _different_ FlutterViewController, and that one should get its own FirstFrameRendered callback, no?

gaaclarke replied:
Sure, it may work.  It wasn't necessary for my test or any existing test so I decided to keep it simple.

dnfield replied:
Does the test still pass without changing this?  If so and we can keep it, it makes it a bit easier for people to keep using the actual app to test something.

gaaclarke replied:
It crashes the test unfortunately:
```
#0	0x00000001012b8d28 in flutter::PersistentCache::ResetStoredNewShaders() [inlined] at /Users/gaaclarke/dev/engine/src/flutter/shell/common/persistent_cache.h:47
#1	0x00000001012b8d28 in flutter::Rasterizer::DoDraw(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >) at /Users/gaaclarke/dev/engine/src/flutter/shell/common/rasterizer.cc:131
#2	0x00000001012b97e4 in decltype(*(std::__1::forward<flutter::Rasterizer*&>(fp0)).*fp(std::__1::forward<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> > >(fp1))) std::__1::__invoke<void (flutter::Rasterizer::*&)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >, void>(void (flutter::Rasterizer::*&)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/type_traits:4291
#3	0x00000001012b97ac in std::__1::__bind_return<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), std::__1::tuple<flutter::Rasterizer*, std::__1::placeholders::__ph<1> >, std::__1::tuple<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&>, __is_valid_bind_return<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), std::__1::tuple<flutter::Rasterizer*, std::__1::placeholders::__ph<1> >, std::__1::tuple<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&> >::value>::type std::__1::__apply_functor<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), std::__1::tuple<flutter::Rasterizer*, std::__1::placeholders::__ph<1> >, 0ul, 1ul, std::__1::tuple<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&> >(void (flutter::Rasterizer::*&)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), std::__1::tuple<flutter::Rasterizer*, std::__1::placeholders::__ph<1> >&, std::__1::__tuple_indices<0ul, 1ul>, std::__1::tuple<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&>&&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:2226
#4	0x00000001012b97ac in std::__1::__bind_return<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), std::__1::tuple<flutter::Rasterizer*, std::__1::placeholders::__ph<1> >, std::__1::tuple<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&>, __is_valid_bind_return<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), std::__1::tuple<flutter::Rasterizer*, std::__1::placeholders::__ph<1> >, std::__1::tuple<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&> >::value>::type std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&>::operator()<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> > >(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:2259
#5	0x00000001012b97ac in decltype(std::__1::forward<std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&>&>(fp)(std::__1::forward<std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> > >(fp0))) std::__1::__invoke<std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&>&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> > >(std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&>&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/type_traits:4350
#6	0x00000001012b97ac in void std::__1::__invoke_void_return_wrapper<void>::__call<std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&>&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> > >(std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&>&, std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/__functional_base:349
#7	0x00000001012b97ac in std::__1::__function::__func<std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&>, std::__1::allocator<std::__1::__bind<void (flutter::Rasterizer::*)(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >), flutter::Rasterizer*, std::__1::placeholders::__ph<1> const&> >, void (std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >)>::operator()(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >&&) at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:1572
#8	0x00000001012b8e74 in std::__1::function<void (std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >)>::operator()(std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >) const [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:1923
#9	0x00000001012b8e5c in flutter::Pipeline<flutter::LayerTree>::Consume(std::__1::function<void (std::__1::unique_ptr<flutter::LayerTree, std::__1::default_delete<flutter::LayerTree> >)>) at /Users/gaaclarke/dev/engine/src/flutter/shell/common/pipeline.h:148
#10	0x00000001012b8b70 in flutter::Rasterizer::Draw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >) at /Users/gaaclarke/dev/engine/src/flutter/shell/common/rasterizer.cc:103
#11	0x00000001012c4e64 in flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25::operator()() const [inlined] at /Users/gaaclarke/dev/engine/src/flutter/shell/common/shell.cc:794
#12	0x00000001012c4e28 in decltype(std::__1::forward<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25&>(fp)()) std::__1::__invoke<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25&>(flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/type_traits:4350
#13	0x00000001012c4e28 in void std::__1::__invoke_void_return_wrapper<void>::__call<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25&>(flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/__functional_base:349
#14	0x00000001012c4e28 in std::__1::__function::__func<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25, std::__1::allocator<flutter::Shell::OnAnimatorDraw(fml::RefPtr<flutter::Pipeline<flutter::LayerTree> >)::$_25>, void ()>::operator()() at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:1572
#15	0x0000000101274c84 in std::__1::function<void ()>::operator()() const [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/functional:1923
#16	0x0000000101274c70 in fml::MessageLoopImpl::FlushTasks(fml::FlushType) at /Users/gaaclarke/dev/engine/src/flutter/fml/message_loop_impl.cc:138
#17	0x0000000101279e30 in fml::MessageLoopDarwin::OnTimerFire(__CFRunLoopTimer*, fml::MessageLoopDarwin*) at /Users/gaaclarke/dev/engine/src/flutter/fml/platform/darwin/message_loop_darwin.mm:75
#18	0x000000018423baa8 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ ()
#19	0x000000018423b76c in __CFRunLoopDoTimer ()
#20	0x000000018423b010 in __CFRunLoopDoTimers ()
#21	0x0000000184238b60 in __CFRunLoopRun ()
#22	0x0000000184158da8 in CFRunLoopRunSpecific ()
#23	0x0000000101279f54 in fml::MessageLoopDarwin::Run() at /Users/gaaclarke/dev/engine/src/flutter/fml/platform/darwin/message_loop_darwin.mm:46
#24	0x0000000101274bc0 in fml::MessageLoopImpl::DoRun() at /Users/gaaclarke/dev/engine/src/flutter/fml/message_loop_impl.cc:84
#25	0x000000010127946c in fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0::operator()() const [inlined] at /Users/gaaclarke/dev/engine/src/flutter/fml/thread.cc:34
#26	0x00000001012793d0 in decltype(std::__1::forward<fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(fp)()) std::__1::__invoke<fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0&&) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/type_traits:4350
#27	0x00000001012793d0 in void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0>&, std::__1::__tuple_indices<>) [inlined] at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/thread:342
#28	0x00000001012793d0 in void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, fml::Thread::Thread(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)::$_0> >(void*) at /Users/gaaclarke/dev/engine/src/buildtools/mac-x64/clang/include/c++/v1/thread:352
```

I think running 2 FlutterEngine's is suspect, I've seen some thing that make my scratch my beard looking through the code.

xster replied:
I think this change is fine. This delegate is akin to a setup phase in a test. We want it to be as undisturbed as possible so we can observe the side effects of loading vms, isolates, engines, entrypoints etc in the actual tests.

dnfield replied:
At some point we'll want this app to actually run Dart code outside of a test though...  

We should be able to run more than one Flutter engine.  We do it in ios_add2app for example.

gaaclarke replied:
You know what, that crash I was seeing some times before I added my test.  Now that I've changed the app delegate, I haven't seen it in my test.  It might not be related to having 2 engines, but having to do with running that view controller from the app delegate for some reason.

dnfield replied:
Ahh ok, well we can figure something out for this.  At some point we'll want to be able to present a FlutterViewController automatically in the app but it's fine for now.

xster replied:
> At some point we'll want this app to actually run Dart code outside of a test though...

Do we? Can you describe it a bit? I expected everything under testing/... to be just bootstrapping for tests.

xster replied:
> We should be able to run more than one Flutter engine. We do it in ios_add2app for example.

We can, but it has side effects (of starting the vm). If we run performance tests against this project, we want to start fresh

dnfield replied:
This isn't a performance test, this is a test that the first frame rendering callback gets called.

See what I'm doing in https://github.com/flutter/engine/pull/10065 - I'd like to do the same for iOS at some point (build the app, upload it to Firebase test lab, get the timeline)

xster replied:
Ya I didn't mean this specific test, just the range of tests this project can support (since we're going off of the idea that this is the first useful test we're writing so we're taking over this project). Since many tests can re-use the same project, I was just making the point above that this current form of the project is going to be useful for many tests.

If you then start writing a test checking for cold full-app launch speed, then that test warrants a new project at that moment.

xster commented on shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm:
```diff
@@ -89,14 +101,19 @@
+    }
+  }
+
+  // Check other bundles if we still haven't found anything.  This is useful for XCTest targets.
```
Can you describe this issue a bit? Where does the framework end up and what does the file structure look like in a test case?

gaaclarke replied:
Sure, it's kind of goofy, here is the file structure of the app that will run:
```
ScenariosUITests-Runner.app
ScenariosUITests-Runner.app/_CodeSignature
ScenariosUITests-Runner.app/_CodeSignature/CodeResources
ScenariosUITests-Runner.app/ScenariosUITests-Runner
ScenariosUITests-Runner.app/PlugIns
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/_CodeSignature
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/_CodeSignature/CodeResources
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/ScenariosUITests
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/Frameworks
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/Frameworks/Flutter.framework
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/Frameworks/Flutter.framework/_CodeSignature
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/Frameworks/Flutter.framework/_CodeSignature/CodeResources
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/Frameworks/Flutter.framework/icudtl.dat
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/Frameworks/Flutter.framework/Flutter
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/Frameworks/Flutter.framework/Info.plist
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/Frameworks/App.framework
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/Frameworks/App.framework/_CodeSignature
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/Frameworks/App.framework/_CodeSignature/CodeResources
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/Frameworks/App.framework/App
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/Frameworks/App.framework/Info.plist
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest/Info.plist
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest.dSYM
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest.dSYM/Contents
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest.dSYM/Contents/Resources
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest.dSYM/Contents/Resources/DWARF
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest.dSYM/Contents/Resources/DWARF/ScenariosUITests
ScenariosUITests-Runner.app/PlugIns/ScenariosUITests.xctest.dSYM/Contents/Info.plist
ScenariosUITests-Runner.app/Frameworks
ScenariosUITests-Runner.app/Frameworks/XCTAutomationSupport.framework
ScenariosUITests-Runner.app/Frameworks/XCTAutomationSupport.framework/_CodeSignature
ScenariosUITests-Runner.app/Frameworks/XCTAutomationSupport.framework/_CodeSignature/CodeResources
ScenariosUITests-Runner.app/Frameworks/XCTAutomationSupport.framework/XPCServices
ScenariosUITests-Runner.app/Frameworks/XCTAutomationSupport.framework/XPCServices/LogArchiveCollector.xpc
ScenariosUITests-Runner.app/Frameworks/XCTAutomationSupport.framework/XPCServices/LogArchiveCollector.xpc/_CodeSignature
ScenariosUITests-Runner.app/Frameworks/XCTAutomationSupport.framework/XPCServices/LogArchiveCollector.xpc/_CodeSignature/CodeResources
ScenariosUITests-Runner.app/Frameworks/XCTAutomationSupport.framework/XPCServices/LogArchiveCollector.xpc/LogArchiveCollector
ScenariosUITests-Runner.app/Frameworks/XCTAutomationSupport.framework/XPCServices/LogArchiveCollector.xpc/Info.plist
ScenariosUITests-Runner.app/Frameworks/XCTAutomationSupport.framework/XPCServices/LogArchiveCollector.xpc/version.plist
ScenariosUITests-Runner.app/Frameworks/XCTAutomationSupport.framework/Info.plist
ScenariosUITests-Runner.app/Frameworks/XCTAutomationSupport.framework/XCTAutomationSupport
ScenariosUITests-Runner.app/Frameworks/XCTAutomationSupport.framework/version.plist
ScenariosUITests-Runner.app/Frameworks/XCTest.framework
ScenariosUITests-Runner.app/Frameworks/XCTest.framework/_CodeSignature
ScenariosUITests-Runner.app/Frameworks/XCTest.framework/_CodeSignature/CodeResources
ScenariosUITests-Runner.app/Frameworks/XCTest.framework/en.lproj
ScenariosUITests-Runner.app/Frameworks/XCTest.framework/en.lproj/InfoPlist.strings
ScenariosUITests-Runner.app/Frameworks/XCTest.framework/XCTest
ScenariosUITests-Runner.app/Frameworks/XCTest.framework/Info.plist
ScenariosUITests-Runner.app/Frameworks/XCTest.framework/version.plist
ScenariosUITests-Runner.app/embedded.mobileprovision
ScenariosUITests-Runner.app/Info.plist
ScenariosUITests-Runner.app/PkgInfo
```

`ScenariosUITests-Runner.app` is the application that gets run, but `Runner.app/PlugIns/ScenariosUITests.xctest` is the target you are building.  So, the code before this change tries to find the compiled app in the main bundle, ie `ScenariosUITests-Runner.app/Frameworks/App.framework`, but there is no way to actually put the framework in there.  So, we are making the code look through it's bundles so it find `Runner.app/PlugIns/ScenariosUITests.xctest/Frameworks/App.framework`

xster replied:
I see. In this case, we should let the test itself ask to push in a custom bundle. From the production side, we should offer APIs to look at different bundles but the production code shouldn't go find them itself. i.e., doesn't FlutterViewController already have a initWithEngine: nibName: bundle: API? If the test knows what its own project structure should be, it should go tell the engine what to load.

gaaclarke replied:
Yep, there is a way to specify the bundle directly.  I can do that if you want.  I don't really see a downside to making our search for the code a bit more robust when it only gets executed if all else fails and the client wasn't already specific.

xster replied:
This would just iterate through their project bundles looking for one named App.framework right? It seems a bit too assumptive vs something explicit. If the users are at a point where their frameworks are sitting in random places in their .app, they probably should know exactly what they're doing.  

@jmagman might have opinions too.

gaaclarke replied:
Started being explicit.

xster commented on testing/scenario_app/ios/Scenarios/Scenarios.xcodeproj/project.pbxproj:
```diff
@@ -471,6 +483,7 @@
 					"$(PROJECT_DIR)",
 				);
 				INFOPLIST_FILE = Scenarios/Info.plist;
+				IPHONEOS_DEPLOYMENT_TARGET = 11.2;
```
Is this a hard restriction?

xster commented on testing/scenario_app/ios/Scenarios/ScenariosTests/FlutterViewControllerTest.m:
```diff
@@ -0,0 +1,51 @@
+//
```
Use our headers rather than the auto-generated one

gaaclarke replied:
Changed it, didn't tweak the template.

xster commented on testing/scenario_app/ios/Scenarios/ScenariosTests/FlutterViewControllerTest.m:
```diff
@@ -0,0 +1,51 @@
+  AppDelegate* appDelegate = (AppDelegate*)UIApplication.sharedApplication.delegate;
+  UIViewController* rootVC = appDelegate.window.rootViewController;
+  [rootVC presentViewController:self.flutterViewController animated:NO completion:nil];
+  NSRunLoop* theRL = [NSRunLoop currentRunLoop];
```
Don't abbreviate

gaaclarke replied:
Removed.

xster commented:
LGTM. I'd prefer if the app.framework finding was explicit (beyond the most default location) but you decide.

jmagman commented on shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm:
```diff
@@ -89,14 +101,19 @@
+
+  // Check other bundles if we still haven't found anything.  This is useful for XCTest targets.
+  if (settings.application_library_path.size() == 0) {
+    for (NSBundle* bundle in [NSBundle allBundles]) {
```
Shouldn't App.framework be one of these bundles, so check for its existence as bundle in the loop instead of checking the Framework directories?  Or is the issue that the test can't load the symbols, but it's still embedded?  Or are there multiple of them, which is why you don't break from the loop when you find the first?  Sorry, I'm not familiar with this code... I would have expected `bundle = [NSBundle bundleWithIdentifier:[FlutterDartProject defaultBundleIdentifier]];` to be the App.framework, even in the tests.

Also nit allBundles is a class method, so dot notation `NSBundle.allBundles`

gaaclarke replied:
I explain why I was doing this in a previous comment to xster.  I've since removed the code so it's moot now.  TLDR App.framework isn't one of the bundles, the .xctest is the bundle that has App.framework in its Frameworks.  That's how we do it elsewhere.

gaaclarke commented on testing/scenario_app/ios/Scenarios/ScenariosTests/FlutterViewControllerTest.m:
```diff
@@ -0,0 +1,51 @@
+
+- (void)setUp {
+  self.continueAfterFailure = NO;
+}
```
Done.

gaaclarke commented on testing/scenario_app/ios/Scenarios/ScenariosTests/FlutterViewControllerTest.m:
```diff
@@ -0,0 +1,51 @@
+}
+
+- (void)tearDown {
+  if (self.flutterViewController) {
```
Done.

jmagman replied:
[super tearDown] should come last in the method, dealloc-style

gaaclarke commented on testing/scenario_app/ios/Scenarios/ScenariosTests/FlutterViewControllerTest.m:
```diff
@@ -0,0 +1,51 @@
+    [theRL runMode:NSDefaultRunLoopMode beforeDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
+    countDownMs -= 100;
+  }
+  XCTAssertGreaterThan(countDownMs, 0);
```
Yep, it works.  You have to do this because the engine is posting things to the main thread that then get posted back to worker threads.

jmagman replied:
This should have been
```
XCTestExpectation *callbackExpecation = [self expectationWithDescription:@"callback"];
```

gaaclarke commented on shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm:
```diff
@@ -25,6 +25,18 @@
 
 static const char* kApplicationKernelSnapshotFileName = "kernel_blob.bin";
 
+static NSString* FindFrameworkApplicationLibraryPath(NSBundle* bundle) {
```
I've removed this diff as per xster's request.

gaaclarke replied:
Removed code as per xsters request.

xster commented:
LGTM

jmagman commented on testing/scenario_app/ios/Scenarios/ScenariosTests/FlutterViewControllerTest.m:
```diff
@@ -0,0 +1,60 @@
+
+static NSBundle* FindTestBundle() {
+  for (NSBundle* bundle in [NSBundle allBundles]) {
+    if ([bundle.bundlePath containsString:@".xctext"]) {
```
Should this be xctest?  If this test is passing, does that prove this isn't needed (because there are no . xctext bundles)?

gaaclarke commented:
I missed jmagman's comments and she was correct that the hunt for the bundle wasn't necessary since I changed the test from a UI test to a standard test.  I missed her comment before this was merged and corrected it here:

https://github.com/flutter/engine/pull/10178
